go/vt/discovery/healthcheck.go

```diff
@@ -467,10 +468,16 @@ func (hc *HealthCheckImpl) deleteTablet(tablet *topodata.Tablet) {
 				continue
 			}
 			delete(ths, tabletAlias)
-			// delete from healthy list
+
 			healthy, ok := hc.healthy[key]
 			if ok && len(healthy) > 0 {
-				hc.recomputeHealthy(key)
+				if tabletType == topodata.TabletType_PRIMARY {
+					// If tablet type is primary, we should only have one tablet in the healthy list.
+					hc.recomputeHealthyPrimary(key)
```

Comment:
This is an interesting bug. We have safeguards in `updateHealth` which make sure that there is exactly one element in `healthy` for the PRIMARY tablet type, but the recomputation here had no such checks.
While the fix seems correct, what do you think about pushing it into `recomputeHealthy` instead of creating a new function call here?
Right now this is the only call site where `recomputeHealthy` is called for the PRIMARY tablet type, but it seems more compact and future proof to change `recomputeHealthy` to maintain the invariant.

Reply:
> While the fix seems correct, what do you think about pushing it into `recomputeHealthy` instead of creating a new function call here?

I feel lik having two separate methods helps highlighting the difference between the recomputation logic for primaries vs any other replica type more clearly.

I was thinking that maybe having a `recomputeHealthy` method that then calls out to either `recomputeHealthyReplicas` and `recomputeHealthyPrimaries` would probably be the best for clarity (and the go compiler should just inline all of these calls anyway, so not a performance issue either).

But then I noticed that pushing down the tablet type check into `recomputeHealth` would require passing down the `key` and `tabletType` as arguments, which seems redundant, because `key` contains the tablet type already, just as a string. Not passing down `key` would require passing down other arguments to be able to build the key, and then things just start to become more messy.

As I mentioned in Slack, I'm very interested giving the healthcheck code a polishing pass separately from this fix, so I'd suggest we punt on this for now. 🙇‍♂️ What do you think?

Reply:
I can see that instead of passing the "key", you would now have to pass the target. When this was written, it just seemed convenient to use the key because it has already been computed. So we have
```
			hc.recomputeHealthy(targetKey)
```
However, later on, we are doing a key computation for the second call
```
			oldTargetKey := KeyFromTarget(prevTarget)
			hc.recomputeHealthy(oldTargetKey)
```
It does not seem too invasive to change `recomputeHealthy` to be
```
func (hc *HealthCheckImpl) recomputeHealthy(target *query.Target) {
			key := KeyFromTarget(target)
			if target.TabletType == ..._PRIMARY {
			...
}
```
I do also see issues with the function name `recomputeHealthyReplicas`.  `replica` is an overloaded term, and it could mean replicas in the sense of replicating from the primary, or it could mean the specific tablet type (as opposed to RDONLY and other tablet types). Refactoring it this way avoids that naming issue and leads to a smaller code change than what you are currently proposing.